### DISCUSSION TITLE
CLI tool to merge TOML files

### DISCUSSION THREAD

**Discussion by halloleo** [toml-lang/toml#1011] [Q&A]
Often I have to merge two TOML files _by table header_. Here an example what I mean by this:

<details>
<summary>file1.toml:</summary>

```
[header_a]
key1 = "information"
key2 = "content"

[header_b]
key1 = "data"
```
</details>

<details>
<summary>file2.toml:</summary>

```
[header_a]
key3 = "details"
```
</details>

<details>
<summary>Merged fileresult.toml:</summary>

```
[header_a]
key1 = "information"
key2 = "content"
key3 = "details"

[header_b]
key1 = "data"
```
</details>

In order to not invent the wheel twice, does a readymade CLI tool for this exist?

**Comment by ChristianSi** [comment]
I'm not aware for an CLI tool – googling found some old stuff, but nothing that looked up-to-date. But it should be fairly easy to write, let's say for your example files in Python (untested code):

```py
import tomllib
import tomli_w

with open("file1.toml", "rb") as fh:
    file1data = tomllib.load(fh)

with open("file2.toml", "rb") as fh:
    file2data = tomllib.load(fh)

fileresult_data = file1data | file2data

with open("fileresult_data.toml", "wb") as fh:
    tomli_w.dump(fileresult_data, fh)
```

Reading the file names from CLI arguments instead is left as exercise for the reader.

**Reply by halloleo** [reply]
Thanks @ChristianSi. 

Makes total sense. I like the union operator on the dictionaries! :-)

**Reply by eksortso** [reply]
@ChristianSi Quick note... Should this be the line defining the result?

```python
fileresult_data = file1data | file2data
```

**Reply by ChristianSi** [reply]
@eksortso Indeed, thanks for spotting that small typo (now corrected!) I guess that's what inevitably happens when you publish untested code...

**Reply by DivineDominion** [reply]
I tried this today, but it didn't produce the expected output:

```sh
cat > a.toml <<- EOF
[header_a]
key1 = "information"
key2 = "content"

[header_b]
key1 = "data"
EOF
```

```sh
cat > b.toml <<- EOF
[header_a]
key3 = "details"
EOF
```

```python
import tomllib
import tomli_w

with open("a.toml", "rb") as fh:
    file1data = tomllib.load(fh)

with open("b.toml", "rb") as fh:
    file2data = tomllib.load(fh)

fileresult_data = file1data | file2data

print(tomli_w.dumps(fileresult_data))
```

Result:

```
[header_a]
key3 = "details"

[header_b]
key1 = "data"
```

Am I holding something wrong?

**Reply by ChristianSi** [reply]
@DivineDominion: Good catch, I had not considered that, while `|` merges dicts, it does so shallowly: if both dicts contain a key, the value from the second one is kept. In our case, however, we want to merge the values, provided they are both dicts (i.e., tables). This code should do the trick:

```py
from typing import Any, Dict

def merge_dicts(a: Dict[Any, Any], b: Dict[Any, Any]) -> Dict[Any, Any]:
    """Merge two dictionaries recursively.

    If both have the same key and their values are dictionaries, merge them recursively.
    Otherwise, b overrides a.
    """
    result = a.copy()
    for key, b_val in b.items():
        a_val = result.get(key)
        if isinstance(a_val, dict) and isinstance(b_val, dict):
            result[key] = merge_dicts(a_val, b_val)
        else:
            result[key] = b_val
    return result

# Read the files first, as above
fileresult_data = merge_dicts(file1data, file2data)
```

**Reply by szepesisz** [reply]
If you don't mind using an external dependency, I would recomment the `deepmerge` library: 
https://github.com/toumorokoshi/deepmerge
It allows the user to define merging strategies depending on the object type.

A came across with it looking for a config merging tool and found out that Adobe-s `himl` (hierarchical yaml) uses this.
https://github.com/adobe/himl
https://github.com/adobe/himl/blob/main/setup.py#L10

**Comment by BatmanAoD** [comment]
The [`tomlq` tool](https://github.com/kislyuk/yq?tab=readme-ov-file#toml-support) included with the Python version of `yq` can do this:

```
tomlq -t -s '.[0] * .[1]' <file1> <file2>
```
